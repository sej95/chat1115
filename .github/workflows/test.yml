name: Test CI

on: [push, pull_request]

jobs:
  test:
    runs-on: ubuntu-latest

    services:
      postgres:
        image: paradedb/paradedb:latest
        env:
          POSTGRES_PASSWORD: postgres
        options: >-
          --health-cmd pg_isready --health-interval 10s --health-timeout 5s  --health-retries 5


        ports:
          - 5432:5432

    steps:
      - uses: actions/checkout@v4

      - name: Install bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: ${{ secrets.BUN_VERSION }}

      - name: Install deps
        run: bun i

      - name: Lint
        run: bun run lint

      - name: Test with Coverage
        run: bun run test:coverage
        env:
          DATABASE_TEST_URL: postgresql://postgres:postgres@localhost:5432/postgres

      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@v4
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: ./coverage/lcov.info
          flags: all
