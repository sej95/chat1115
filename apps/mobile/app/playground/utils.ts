import { ComponentItem } from './type';

/**
 * 组件扫描器 - 用于自动发现和配置组件信息
 */

// 手动配置的组件列表，包含详细信息
export const COMPONENT_CONFIGS: ComponentItem[] = [
  {
    category: 'feedback',
    description: '功能完整的提示框组件，支持12种位置和多种触发方式',
    hasDemos: true,
    hasReadme: true,
    name: 'Tooltip',
    path: 'tooltip',
    tags: ['tooltip', 'popover', 'overlay'],
  },
  {
    category: 'feedback',
    description: '优雅的消息提示组件，支持多种类型和动画效果',
    hasDemos: true,
    hasReadme: true,
    name: 'Toast',
    path: 'toast',
    tags: ['toast', 'notification', 'message'],
  },
  {
    category: 'display',
    description: '基于 Shiki 的高性能代码高亮组件，支持100+种编程语言',
    hasDemos: true,
    hasReadme: true,
    name: 'Highlighter',
    path: 'highlighter',
    tags: ['code', 'syntax', 'highlight', 'shiki'],
  },
  {
    category: 'display',
    description: '强大的Markdown渲染组件，支持数学公式、代码高亮等',
    hasDemos: true,
    hasReadme: true,
    name: 'Markdown',
    path: 'markdown',
    tags: ['markdown', 'text', 'render', 'mathjax', 'code'],
  },
  {
    category: 'layout',
    description: '功能丰富的列表项组件，支持头像、描述、路由导航',
    hasDemos: true,
    hasReadme: true,
    name: 'ListItem',
    path: 'listitem',
    tags: ['list', 'item', 'avatar', 'navigation', 'menu'],
  },
  {
    category: 'display',
    description: '可定制的头像组件，支持自定义尺寸、边框和错误处理',
    hasDemos: true,
    hasReadme: true,
    name: 'Avatar',
    path: 'avatar',
    tags: ['avatar', 'user', 'profile', 'image'],
  },
  {
    category: 'layout',
    description: '设置组件之间的间距，支持水平/垂直布局、不同对齐方式和分隔符',
    hasDemos: true,
    hasReadme: true,
    name: 'Space',
    path: 'space',
    tags: ['space', 'layout', 'gap', 'margin', 'separator'],
  },
  {
    category: 'display',
    description: '微软 Fluent 风格的 3D 表情符号组件，支持自定义大小和回退',
    hasDemos: true,
    hasReadme: false,
    name: 'FluentEmoji',
    path: 'fluentemoji',
    tags: ['emoji', 'fluent', '3d', 'icon'],
  },
  {
    category: 'display',
    description: '进行标记和分类的小标签组件，支持自定义样式',
    hasDemos: true,
    hasReadme: false,
    name: 'Tag',
    path: 'tag',
    tags: ['tag', 'label', 'badge', 'category'],
  },
  {
    category: 'navigation',
    description: '水平滚动的胶囊选项卡组件，支持自定义样式和选择状态',
    hasDemos: true,
    hasReadme: false,
    name: 'CapsuleTabs',
    path: 'capsuletabs',
    tags: ['tabs', 'capsule', 'navigation', 'scroll', 'horizontal'],
  },
  {
    category: 'basic',
    description: 'React Native版本的按钮组件，参考Ant Design设计，支持多种类型、尺寸和状态',
    hasDemos: true,
    hasReadme: false,
    name: 'Button',
    path: 'button',
    tags: ['button', 'action', 'click', 'primary', 'loading'],
  },
  {
    category: 'feedback',
    description: '骨架屏组件，用于页面加载状态，支持头像、标题、段落和动画效果',
    hasDemos: true,
    hasReadme: false,
    name: 'Skeleton',
    path: 'skeleton',
    tags: ['skeleton', 'loading', 'placeholder', 'shimmer', 'animation'],
  },
  {
    category: 'form',
    description: '支持异步操作的即时开关组件，具有乐观更新和错误回滚机制',
    hasDemos: true,
    hasReadme: true,
    name: 'InstantSwitch',
    path: 'instant-switch',
    tags: ['switch', 'toggle', 'async', 'loading', 'form'],
  },
  {
    category: 'form',
    description: '基于 LobeUI 重写的颜色选择器组件，支持圆形/方形样式、透明色处理',
    hasDemos: true,
    hasReadme: true,
    name: 'ColorSwatches',
    path: 'colorswatches',
    tags: ['color', 'picker', 'swatch', 'form', 'ui'],
  },
  {
    category: 'display',
    description: '基于 LobeUI 重写的色板展示组件，支持完整的颜色级别、透明色和复制功能',
    hasDemos: true,
    hasReadme: true,
    name: 'ColorScales',
    path: 'colorscales',
    tags: ['color', 'scale', 'palette', 'design', 'theme'],
  },
  {
    category: 'display',
    description: '主题 Token 展示页面，显示种子、映射和别名令牌，支持主题切换和搜索',
    hasDemos: false,
    hasReadme: false,
    name: 'Theme Token',
    path: 'theme-token',
    tags: ['theme', 'tokens', 'design', 'colors', 'typography'],
  },
  {
    category: 'display',
    description: '主题展示页面，如何使用 ThemeProvider',
    hasDemos: false,
    hasReadme: false,
    name: 'ThemeProvider',
    path: 'theme-provider',
    tags: ['theme', 'token'],
  },
];

/**
 * 根据分类获取组件列表
 */
export const getComponentsByCategory = (category?: string): ComponentItem[] => {
  if (!category) {
    return COMPONENT_CONFIGS;
  }
  return COMPONENT_CONFIGS.filter((component) => component.category === category);
};

/**
 * 根据标签搜索组件
 */
export const searchComponentsByTag = (tag: string): ComponentItem[] => {
  return COMPONENT_CONFIGS.filter((component) =>
    component.tags.some((t: string) => t.toLowerCase().includes(tag.toLowerCase())),
  );
};

/**
 * 根据名称搜索组件
 */
export const searchComponentsByName = (name: string): ComponentItem[] => {
  return COMPONENT_CONFIGS.filter(
    (component) =>
      component.name.toLowerCase().includes(name.toLowerCase()) ||
      component.description.toLowerCase().includes(name.toLowerCase()),
  );
};

/**
 * 获取所有分类
 */
export const getAllCategories = (): string[] => {
  const categories = new Set(COMPONENT_CONFIGS.map((component) => component.category));
  return Array.from(categories);
};

/**
 * 获取所有标签
 */
export const getAllTags = (): string[] => {
  const tags = new Set(COMPONENT_CONFIGS.flatMap((component) => component.tags));
  return Array.from(tags);
};

/**
 * 获取组件统计信息
 */
export const getComponentStats = () => {
  const total = COMPONENT_CONFIGS.length;
  const withReadme = COMPONENT_CONFIGS.filter((c) => c.hasReadme).length;
  const withDemos = COMPONENT_CONFIGS.filter((c) => c.hasDemos).length;

  return {
    categories: getAllCategories().length,
    tags: getAllTags().length,
    total,
    withDemos,
    withReadme,
  };
};
