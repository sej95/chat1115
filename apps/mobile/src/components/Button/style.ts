import { createStyles } from '@/theme';

export type ButtonType = 'primary' | 'default' | 'text' | 'link';
export type ButtonSize = 'small' | 'middle' | 'large';

export const useStyles = createStyles(
  (
    token,
    {
      type,
      size,
      disabled,
      block,
    }: { block: boolean; disabled: boolean; size: ButtonSize; type: ButtonType },
  ) => {
    const getTypeStyles = () => {
      switch (type) {
        case 'primary': {
          return {
            backgroundColor: disabled ? token.colorBgContainerDisabled : token.colorPrimary,
            borderColor: disabled ? token.colorPrimaryBorder : token.colorBgElevated,
            textColor: disabled ? token.colorTextDisabled : token.colorBgLayout,
          };
        }
        case 'text': {
          return {
            backgroundColor: 'transparent',
            borderColor: 'transparent',
            textColor: disabled ? token.colorTextDisabled : token.colorPrimary,
          };
        }
        case 'link': {
          return {
            backgroundColor: 'transparent',
            borderColor: 'transparent',
            textColor: disabled ? token.colorTextDisabled : token.colorPrimary,
          };
        }
        default: {
          return {
            backgroundColor: disabled ? token.colorBgContainerDisabled : token.colorBgContainer,
            borderColor: disabled ? token.colorBorder : token.colorBorder,
            textColor: disabled ? token.colorTextDisabled : token.colorText,
          };
        }
      }
    };

    const getSizeStyles = () => {
      switch (size) {
        case 'small': {
          return {
            fontSize: token.fontSizeSM,
            minHeight: 24,
            paddingHorizontal: token.paddingXS,
            paddingVertical: token.paddingXXS,
          };
        }
        case 'large': {
          return {
            fontSize: token.fontSizeLG,
            minHeight: 40,
            paddingHorizontal: token.paddingLG,
            paddingVertical: token.paddingSM,
          };
        }
        default: {
          return {
            fontSize: token.fontSize,
            minHeight: 32,
            paddingHorizontal: token.paddingSM,
            paddingVertical: token.paddingXS,
          };
        }
      }
    };

    const typeStyles = getTypeStyles();
    const sizeStyles = getSizeStyles();

    return {
      button: {
        alignItems: 'center' as const,
        backgroundColor: typeStyles.backgroundColor,
        borderColor: typeStyles.borderColor,
        borderRadius: token.borderRadius,
        borderWidth: type === 'text' || type === 'link' ? 0 : 1,
        flexDirection: 'row' as const,
        justifyContent: 'center' as const,
        minHeight: sizeStyles.minHeight,
        opacity: disabled ? 0.6 : 1,
        paddingHorizontal: sizeStyles.paddingHorizontal,
        paddingVertical: sizeStyles.paddingVertical,
        width: block ? '100%' : 'auto',
      },
      loading: {
        marginRight: token.marginXS,
      },
      text: {
        color: typeStyles.textColor,
        fontSize: sizeStyles.fontSize,
        fontWeight: type === 'primary' ? '600' : '400',
        textAlign: 'center' as const,
        textDecorationLine: type === 'link' ? 'underline' : 'none',
      },
    };
  },
);
