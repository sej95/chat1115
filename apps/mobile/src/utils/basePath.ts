// Mobile端不需要复杂的basePath处理，直接返回原始路径
export const withBasePath = (path: string): string => {
  return `http://localhost:3020${path}`;
};
