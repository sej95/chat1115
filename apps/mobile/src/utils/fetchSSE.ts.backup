import { DEFAULT_MODEL } from '@/const/settings';
import EventSource from 'react-native-sse';

const START_ANIMATION_SPEED = 4;
const END_ANIMATION_SPEED = 15;

interface SmoothMessageParams {
  onTextUpdate: (delta: string, text: string) => void;
  startSpeed?: number;
}

const createSmoothMessage = (params: SmoothMessageParams) => {
  const { startSpeed = START_ANIMATION_SPEED } = params;

  let buffer = '';
  // why use queue: https://shareg.pt/GLBrjpK
  const outputQueue: string[] = [];
  let isAnimationActive = false;
  let animationFrameId: number | null = null;

  // when you need to stop the animation, call this function
  const stopAnimation = () => {
    isAnimationActive = false;
    if (animationFrameId !== null) {
      cancelAnimationFrame(animationFrameId);
      animationFrameId = null;
    }
  };

  // define startAnimation function to display the text in buffer smooth
  // when you need to start the animation, call this function
  const startAnimation = (speed = startSpeed) =>
    new Promise<void>((resolve) => {
      if (isAnimationActive) {
        resolve();
        return;
      }

      isAnimationActive = true;

      const updateText = () => {
        // 如果动画已经不再激活，则停止更新文本
        if (!isAnimationActive) {
          if (animationFrameId) cancelAnimationFrame(animationFrameId);
          animationFrameId = null;
          resolve();
          return;
        }

        // 如果还有文本没有显示
        // 检查队列中是否有字符待显示
        if (outputQueue.length > 0) {
          // 从队列中获取前 n 个字符（如果存在）
          const charsToAdd = outputQueue.splice(0, speed).join('');
          buffer += charsToAdd;

          // 更新消息内容，这里可能需要结合实际情况调整
          params.onTextUpdate(charsToAdd, buffer);
        } else {
          // 当所有字符都显示完毕时，清除动画状态
          isAnimationActive = false;
          animationFrameId = null;
          resolve();
          return;
        }

        animationFrameId = requestAnimationFrame(updateText);
      };

      animationFrameId = requestAnimationFrame(updateText);
    });

  const pushToQueue = (text: string) => {
    outputQueue.push(...text.split(''));
  };

  return {
    getBuffer: () => buffer,
    isAnimationActive,
    isTokenRemain: () => outputQueue.length > 0,
    pushToQueue,
    startAnimation,
    stopAnimation,
  };
};

interface FetchSSEParams {
  apiKey: string;
  messages: {
    content: string;
    role: string;
  }[];
  onDone: () => void;
  onError: (error: any) => void;
  onMessage: (content: string) => void;
  proxy: string;
}

export const fetchSSE = ({
  proxy,
  apiKey,
  messages,
  onMessage,
  onDone,
  onError,
}: FetchSSEParams) => {
  const smoothMessageController = createSmoothMessage({
    onTextUpdate: (delta, text) => {
      onMessage(text);
    },
  });

  const es = new EventSource(`${proxy}/v1/chat/completions`, {
    body: JSON.stringify({
      messages: messages.map(({ role, content }) => ({
        content,
        role,
      })),
      model: DEFAULT_MODEL,
      stream: true,
    }),
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    method: 'POST',
  });

  // 处理消息事件
  es.addEventListener('message', (event) => {
    try {
      const rawData = event.data;
      if (typeof rawData !== 'string') {
        return;
      }

      if (rawData.trim() === '[DONE]') {
        smoothMessageController.stopAnimation();
        if (smoothMessageController.isTokenRemain()) {
          smoothMessageController.startAnimation(END_ANIMATION_SPEED).then(() => {
            onDone();
          });
        } else {
          onDone();
        }
        es.close();
        return;
      }

      const data = JSON.parse(rawData);
      const chunkContent = data.choices?.[0]?.delta?.content || '';
      if (chunkContent) {
        smoothMessageController.pushToQueue(chunkContent);
        if (!smoothMessageController.isAnimationActive) {
          smoothMessageController.startAnimation();
        }
      }
    } catch (err) {
      console.error('解析消息失败:', err, '原始数据:', event.data);
      smoothMessageController.stopAnimation();
      onError(err);
      es.close();
    }
  });

  // 处理错误事件
  es.addEventListener('error', (error) => {
    console.error('SSE 错误:', error);
    smoothMessageController.stopAnimation();
    onError(error);
    if (es && typeof es.close === 'function') {
      es.close();
    }
  });

  return {
    close: () => {
      smoothMessageController.stopAnimation();
      es.close();
    },
  };
};
