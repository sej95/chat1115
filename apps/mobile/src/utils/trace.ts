import { LOBE_CHAT_TRACE_HEADER, LOBE_CHAT_TRACE_ID, TracePayload } from '@/const/trace';

export const getTracePayload = (req: Request): TracePayload | undefined => {
  const header = req.headers.get(LOBE_CHAT_TRACE_HEADER);
  if (!header) return;

  // Use atob for base64 decoding in React Native environment
  return JSON.parse(atob(header));
};

export const getTraceId = (res: Response) => res.headers.get(LOBE_CHAT_TRACE_ID);

const createTracePayload = (data: TracePayload) => {
  // Use btoa for base64 encoding in React Native environment
  return btoa(JSON.stringify(data));
};

export const createTraceHeader = (data: TracePayload) => {
  return { [LOBE_CHAT_TRACE_HEADER]: createTracePayload(data) };
};
