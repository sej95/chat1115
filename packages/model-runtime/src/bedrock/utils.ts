/**
 * Utility functions for Bedrock model runtime
 */

// Re-export from dedicated utility
export { parseModelList, validateModelListSyntax } from './modelListParser';

// Type imports
type RequestInit = globalThis.RequestInit;

/**
 * Creates HTTP request options with timeout for Bedrock API calls
 * @param body - Request body
 * @param bearerToken - Authorization token
 * @param signal - Optional abort signal
 * @returns Fetch options and cleanup function
 */
export function createBedrockRequestOptions(
  body: any,
  bearerToken: string,
  signal?: AbortSignal,
): { cleanup: () => void; options: RequestInit } {
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), 30_000); // 30s timeout

  return {
    cleanup: () => clearTimeout(timeoutId),
    options: {
      body: JSON.stringify(body),
      headers: {
        'Authorization': `Bearer ${bearerToken}`,
        'Content-Type': 'application/json',
      },
      method: 'POST',
      signal: signal || controller.signal,
    },
  };
}

/**
 * Executes a fetch request with retry logic
 * @param url - Request URL
 * @param options - Fetch options
 * @param maxRetries - Maximum number of retries (default: 3)
 * @returns Promise<Response>
 */
export async function fetchWithRetry(
  url: string,
  options: RequestInit,
  maxRetries = 3,
): Promise<Response> {
  let lastError: Error;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      const response = await fetch(url, options);

      // Don't retry on client errors (4xx) except 429 (rate limit)
      if (response.status >= 400 && response.status < 500 && response.status !== 429) {
        return response;
      }

      // Retry on server errors (5xx) or rate limits (429)
      if (response.status >= 500 || response.status === 429) {
        if (attempt === maxRetries) return response;
        await new Promise((resolve) => {
          setTimeout(resolve, Math.pow(2, attempt) * 1000);
        });
        continue;
      }

      return response;
    } catch (error) {
      lastError = error as Error;
      if (attempt === maxRetries) break;
      await new Promise((resolve) => {
        setTimeout(resolve, Math.pow(2, attempt) * 1000);
      });
    }
  }

  throw lastError!;
}

/**
 * Regex for reliable Claude model detection
 */
export const CLAUDE_MODEL_REGEX = /^claude[\w-]*(:|$)/i;

/**
 * Builds HTTP request body for Bedrock Converse API
 * @param payload - Chat stream payload
 * @param systemMessage - Optional system message
 * @param converseMessages - Processed user/assistant messages
 * @returns Request body object
 */
export function buildConverseRequestBody(
  payload: {
    max_tokens?: number;
    model: string;
    temperature?: number;
    tools?: any[];
    top_p?: number;
  },
  converseMessages: any[],
  systemMessage?: { content: string },
  tools?: any[],
): any {
  const requestBody: any = {
    inferenceConfig: {
      maxTokens: payload.max_tokens || 4096,
    },
    messages: converseMessages,
  };

  // Add system message if present
  if (systemMessage?.content) {
    requestBody.system = [{ text: systemMessage.content }];
  }

  // Add temperature with model-specific scaling
  if (payload.temperature !== undefined) {
    if (CLAUDE_MODEL_REGEX.test(payload.model)) {
      requestBody.inferenceConfig.temperature = payload.temperature / 2;
    } else {
      requestBody.inferenceConfig.temperature = payload.temperature;
    }
  }

  if (payload.top_p !== undefined) {
    requestBody.inferenceConfig.topP = payload.top_p;
  }

  // Add tools if present
  if (tools && tools.length > 0) {
    requestBody.toolConfig = { tools };
  }

  return requestBody;
}

/**
 * Processes and validates chat messages for Bedrock Converse API
 * @param messages - Raw chat messages
 * @returns Processed messages and system message
 */
export function processConverseMessages(messages: any[]): {
  converseMessages: any[];
  systemMessage?: { content: string };
} {
  const systemMessage = messages.find(
    (m): m is { content: string; role: 'system' } =>
      m.role === 'system' && typeof m.content === 'string',
  );

  const userMessages = messages.filter((m) => m.role !== 'system');

  const converseMessages = userMessages
    .filter((msg) => msg.content && (msg.content as string).trim())
    .map((msg) => {
      let mappedRole: string;
      switch (msg.role) {
        case 'assistant': {
          mappedRole = 'assistant';
          break;
        }
        case 'user': {
          mappedRole = 'user';
          break;
        }
        case 'system': {
          mappedRole = 'system';
          break;
        }
        default: {
          console.warn(`Unknown message role encountered: ${msg.role}. Defaulting to 'user'.`);
          mappedRole = 'user';
        }
      }
      return {
        content: [{ text: (msg.content as string).trim() }],
        role: mappedRole,
      };
    });

  return { converseMessages, systemMessage };
}

/**
 * Valid AWS Bedrock regions
 */
export const VALID_BEDROCK_REGIONS = new Set([
  'us-east-1',
  'us-east-2',
  'us-west-2',
  'ap-south-1',
  'ap-south-2',
  'ap-northeast-1',
  'ap-northeast-2',
  'ap-northeast-3',
  'ap-southeast-1',
  'ap-southeast-2',
  'ca-central-1',
  'eu-central-1',
  'eu-central-2',
  'eu-west-1',
  'eu-west-2',
  'eu-west-3',
  'eu-north-1',
  'eu-south-1',
  'eu-south-2',
  'sa-east-1',
  'us-gov-east-1',
  'us-gov-west-1',
]);

/**
 * Handles Bedrock API errors and creates appropriate error responses
 * @param error - The caught error
 * @param region - AWS region
 * @returns AgentRuntimeError
 */
export function handleBedrockError(error: any, region: string): any {
  // Import dynamically to avoid circular dependencies
  const { AgentRuntimeError } = require('../utils/createError');
  const { AgentRuntimeErrorType } = require('../error');
  const { ModelProvider } = require('../types');

  return AgentRuntimeError.chat({
    error: {
      body: undefined,
      message: error.message,
      type: error.name,
    },
    errorType: AgentRuntimeErrorType.ProviderBizError,
    provider: ModelProvider.Bedrock,
    region,
  });
}
