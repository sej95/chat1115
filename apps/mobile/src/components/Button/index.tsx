import React from 'react';
import {
  TouchableOpacity,
  Text,
  ActivityIndicator,
  TouchableOpacityProps,
  TextStyle,
  ViewStyle,
} from 'react-native';

import { useStyles, ButtonType, ButtonSize } from './style';

export interface ButtonProps extends Omit<TouchableOpacityProps, 'style'> {
  block?: boolean;
  children?: React.ReactNode;
  disabled?: boolean;
  loading?: boolean;
  onPress?: () => void;
  size?: ButtonSize;
  style?: ViewStyle;
  textStyle?: TextStyle;
  type?: ButtonType;
}

const Button: React.FC<ButtonProps> = ({
  type = 'default',
  size = 'middle',
  loading = false,
  disabled = false,
  block = false,
  children,
  onPress,
  style,
  textStyle,
  ...rest
}) => {
  const { styles } = useStyles({ block, disabled: disabled || loading, size, type });

  const handlePress = () => {
    if (!disabled && !loading && onPress) {
      onPress();
    }
  };

  return (
    <TouchableOpacity
      {...rest}
      accessibilityRole="button"
      activeOpacity={disabled || loading ? 1 : 0.7}
      disabled={disabled || loading}
      onPress={handlePress}
      style={[styles.button, style]}
      testID="button"
    >
      {loading && (
        <ActivityIndicator
          color={styles.text.color}
          size="small"
          style={styles.loading}
          testID="loading-indicator"
        />
      )}
      <Text style={[styles.text, textStyle]}>{children}</Text>
    </TouchableOpacity>
  );
};

export default Button;
