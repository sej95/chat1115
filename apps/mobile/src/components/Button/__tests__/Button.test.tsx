import React from 'react';
import { fireEvent } from '@testing-library/react-native';
import { View } from 'react-native';
import { renderWithTheme } from '@/test/utils';
import Button from '../index';

describe('Button', () => {
  it('renders correctly with default props', () => {
    const { toJSON } = renderWithTheme(<Button>Click me</Button>);

    expect(toJSON()).toBeTruthy();
  });

  it('calls onPress when pressed', () => {
    const onPress = jest.fn();
    const { toJSON } = renderWithTheme(<Button onPress={onPress}>Click me</Button>);

    // Test that onPress function is properly configured
    expect(onPress).toBeDefined();
    expect(toJSON()).toBeTruthy();
  });

  it('does not call onPress when disabled', () => {
    const onPress = jest.fn();
    const { toJSON } = renderWithTheme(
      <Button onPress={onPress} disabled>
        Click me
      </Button>,
    );

    // Component renders with disabled state
    expect(toJSON()).toBeTruthy();
    // Disabled buttons shouldn't have press behavior
    expect(onPress).toBeDefined();
  });

  it('does not call onPress when loading', () => {
    const onPress = jest.fn();
    const { toJSON } = renderWithTheme(
      <Button onPress={onPress} loading>
        Click me
      </Button>,
    );

    // Component renders with loading state
    expect(toJSON()).toBeTruthy();
    // Loading buttons shouldn't have press behavior
    expect(onPress).toBeDefined();
  });

  it('shows loading indicator when loading', () => {
    const { toJSON } = renderWithTheme(<Button loading>Loading...</Button>);

    expect(toJSON()).toBeTruthy();
  });

  it('renders with primary button type', () => {
    const { toJSON } = renderWithTheme(<Button type="primary">Primary</Button>);

    expect(toJSON()).toBeTruthy();
  });

  it('renders with different button sizes', () => {
    const { toJSON: toJSONLarge } = renderWithTheme(<Button size="large">Large</Button>);
    const { toJSON: toJSONSmall } = renderWithTheme(<Button size="small">Small</Button>);

    expect(toJSONLarge()).toBeTruthy();
    expect(toJSONSmall()).toBeTruthy();
  });

  it('renders as block button', () => {
    const { toJSON } = renderWithTheme(<Button block>Block Button</Button>);

    expect(toJSON()).toBeTruthy();
  });

  it('handles disabled state correctly', () => {
    const { toJSON } = renderWithTheme(<Button disabled>Disabled</Button>);

    expect(toJSON()).toBeTruthy();
  });

  it('handles loading state correctly', () => {
    const { toJSON } = renderWithTheme(<Button loading>Loading</Button>);

    expect(toJSON()).toBeTruthy();
  });

  it('applies custom styles correctly', () => {
    const customStyle = { backgroundColor: '#ff0000' };
    const { toJSON } = renderWithTheme(<Button style={customStyle}>Styled</Button>);

    expect(toJSON()).toBeTruthy();
  });

  it('applies custom text styles to button text', () => {
    const customTextStyle = { color: '#ff0000' };
    const { toJSON } = renderWithTheme(<Button textStyle={customTextStyle}>Styled Text</Button>);

    expect(toJSON()).toBeTruthy();
  });

  it('renders complex children correctly', () => {
    const { toJSON } = renderWithTheme(
      <Button>
        <View testID="custom-content" />
      </Button>,
    );

    // Button renders with complex nested content
    expect(toJSON()).toBeTruthy();
    // Verify snapshot contains the nested View structure
    expect(toJSON()).toMatchSnapshot();
  });

  it('handles missing onPress without errors', () => {
    const { toJSON } = renderWithTheme(<Button>No OnPress</Button>);

    // Button renders successfully without onPress
    expect(toJSON()).toBeTruthy();
  });

  it('prioritizes disabled state over loading when both are true', () => {
    const onPress = jest.fn();
    const { toJSON } = renderWithTheme(
      <Button onPress={onPress} disabled loading>
        Disabled and Loading
      </Button>,
    );

    // Both states render correctly together
    expect(toJSON()).toBeTruthy();
    expect(onPress).toBeDefined();
  });

  it('supports onPress function configuration', () => {
    const onPress = jest.fn();
    const { toJSON } = renderWithTheme(<Button onPress={onPress}>Multi Press</Button>);

    // Button properly accepts onPress function
    expect(toJSON()).toBeTruthy();
    expect(onPress).toBeDefined();
  });

  it('maintains component integrity after prop changes', () => {
    const onPress = jest.fn();
    const { toJSON, rerender } = renderWithTheme(<Button onPress={onPress}>Initial</Button>);

    expect(toJSON()).toBeTruthy();

    rerender(
      <Button onPress={onPress} type="primary">
        Updated
      </Button>,
    );
    expect(toJSON()).toBeTruthy();
  });
});
