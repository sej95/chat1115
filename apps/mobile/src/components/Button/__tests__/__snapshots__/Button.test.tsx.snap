// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders complex children correctly 1`] = `
<button
  className="css-view-175oi2r r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-cursor-1loqt21 r-touchAction-1otgn73"
  data-testid="button"
  dir={null}
  onClick={[Function]}
  onContextMenu={[Function]}
  onKeyDown={[Function]}
  role="button"
  style={
    {
      "alignItems": "center",
      "backgroundColor": "rgba(255,255,255,1.00)",
      "borderBottomColor": "rgba(217,217,217,1.00)",
      "borderBottomLeftRadius": "8px",
      "borderBottomRightRadius": "8px",
      "borderBottomWidth": "1px",
      "borderLeftColor": "rgba(217,217,217,1.00)",
      "borderLeftWidth": "1px",
      "borderRightColor": "rgba(217,217,217,1.00)",
      "borderRightWidth": "1px",
      "borderTopColor": "rgba(217,217,217,1.00)",
      "borderTopLeftRadius": "8px",
      "borderTopRightRadius": "8px",
      "borderTopWidth": "1px",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": "32px",
      "opacity": 1,
      "paddingBottom": "8px",
      "paddingLeft": "12px",
      "paddingRight": "12px",
      "paddingTop": "8px",
      "transitionDuration": "0s",
      "width": "auto",
    }
  }
  type="button"
>
  <div
    className="css-text-146c3p1"
    dir="auto"
    style={
      {
        "color": "rgba(0,0,0,1.00)",
        "fontWeight": "400",
        "textAlign": "center",
        "textDecoration": "none",
      }
    }
  >
    <div
      className="css-view-175oi2r r-display-xoduu5"
      data-testid="custom-content"
      dir={null}
    />
  </div>
</button>
`;
